File: 2021/src/bin/03.rs
//! Day 3: Binary Diagnostic

use aoc2021::Input;
use itertools::partition;
use std::io::{Error, ErrorKind, Result};

fn main() -> Result<()> {
    let infile: String = match std::env::args().nth(1) {
        Some(v) => v,
        None => {
            return Err(Error::new(
                ErrorKind::InvalidInput,
                "no input file specified",
            ));
        }
    };
    let input = Input::from_file(infile)?;

    let values: Vec<usize> = input
        .lines()
        .unwrap()
        .iter()
        .map(|s| usize::from_str_radix(s, 2).unwrap())
        .collect();

    // Part 1: Power consumption.
    println!("power consumption = {}", power_consumption(&values));

    // Part 2: Life support rating.
    println!("life support rating = {}", life_support_rating(&values));

    Ok(())
}

fn power_consumption(readings: &[usize]) -> usize {
    let gamma = gamma_rate(readings);
    let epsilon = epsilon_rate(gamma);
    gamma * epsilon
}

fn gamma_rate(readings: &[usize]) -> usize {
    let num_readings = readings.len();
    let mut gamma = 0;
    for i in 0..12 {
        let mask = 1 << i;
        let mut ones = 0;
        for r in readings {
            if r & mask == mask {
                ones += 1;
            }
        }
        if ones >= num_readings / 2 {
            gamma |= mask;
        }
    }
    gamma
}

fn epsilon_rate(gamma: usize) -> usize {
    !gamma & 0b1111_1111_1111
}

fn life_support_rating(readings: &[usize]) -> usize {
    let oxygen = oxygen_generator_rating(readings);
    let co2_scrub = co2_scrubber_rating(readings);
    oxygen * co2_scrub
}

fn oxygen_generator_rating(readings: &[usize]) -> usize {
    let mut readings: Vec<usize> = readings.to_vec();
    for i in (0..12).rev() {
        let mask = 1 << i;
        let split_index = partition(&mut readings, |&v| v & mask == mask);
        readings = if split_index >= readings.len() / 2 {
            readings[0..split_index].to_vec()
        } else {
            readings[split_index..].to_vec()
        };
        if readings.len() == 1 {
            break;
        }
    }
    readings[0]
}

fn co2_scrubber_rating(readings: &[usize]) -> usize {
    let mut readings: Vec<usize> = readings.to_vec();
    for i in (0..12).rev() {
        let mask = 1 << i;
        let split_index = partition(&mut readings, |&v| v & mask == mask);
        readings = if split_index < readings.len() / 2 {
            readings[0..split_index].to_vec()
        } else {
            readings[split_index..].to_vec()
        };
        if readings.len() == 1 {
            break;
        }
    }
    readings[0]
}

File: 2021/03.go
package main

import (
	"bufio"
	_ "embed"
	"fmt"
	"os"
	"strconv"
	"strings"
)

//go:embed input/03
var input string

func main() {
	var (
		readings []uint
		scanner  = bufio.NewScanner(strings.NewReader(input))
	)
	for scanner.Scan() {
		if scanner.Text() == "" {
			continue
		}

		n, err := strconv.ParseUint(scanner.Text(), 2, 16)
		if err != nil {
			fmt.Fprintf(os.Stderr, "parse error: %q: %s\n", scanner.Text(), err)
			os.Exit(1)
		}

		readings = append(readings, uint(n))
	}

	// Part 1: Power consumption.
	fmt.Println(powerConsumption(readings))

	// Part 2: Life support rating.
	fmt.Println(lifeSupportRating(readings))
}

func powerConsumption(readings []uint) uint {
	// Calculate the "gamma rate".
	var gamma uint
	for i := 0; i < 12; i++ {
		var (
			mask = uint(1 << i)
			ones int
		)
		for _, r := range readings {
			if r&mask == mask {
				ones++
			}
		}
		if ones >= len(readings)/2 {
			gamma |= mask
		}
	}

	// Calculate the "epsilon rate".
	epsilon := ^gamma & 0b111111111111

	// Return the power consumption.
	return gamma * epsilon
}

func lifeSupportRating(readings []uint) uint {
	return oxygenRating(readings) * co2ScrubberRating(readings)
}

func oxygenRating(readings []uint) uint {
	rs := make([]uint, len(readings))
	copy(rs, readings)

	for i := 11; i >= 0; i-- {
		mask := uint(1 << i)
		split := partition(rs, func(n uint) bool { return n&mask == mask })
		if split >= len(rs)/2 {
			rs = rs[:split]
		} else {
			rs = rs[split:]
		}
		if len(rs) == 1 {
			break
		}
	}
	return rs[0]
}

// partition sorts readings so that elements for which the predicate function
// return true, are earlier in the slice.
// partition returns the index for the first value that does not match the
// predicate.
// This function does not maintain the ordering of elements after the split.
func partition(readings []uint, predicate func(uint) bool) int {
	split := 0
	for i := 0; i < len(readings); i++ {
		if predicate(readings[i]) {
			readings[split], readings[i] = readings[i], readings[split]
			split++
		}
	}
	return split
}

func co2ScrubberRating(readings []uint) uint {
	rs := make([]uint, len(readings))
	copy(rs, readings)

	for i := 11; i >= 0; i-- {
		mask := uint(1 << i)
		split := partition(rs, func(n uint) bool { return n&mask == mask })
		if split < len(rs)/2 {
			rs = rs[:split]
		} else {
			rs = rs[split:]
		}
		if len(rs) == 1 {
			break
		}
	}
	return rs[0]
}
